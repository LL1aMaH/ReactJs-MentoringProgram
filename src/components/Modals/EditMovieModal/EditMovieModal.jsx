import React, { useCallback, useState } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import { Formik, Form } from 'formik';

import { InputField } from '../../InputField/InputField';
import { Button } from '../../Button/Button';
import { Modal } from '../../Modal/Modal';

import Close from '../../../../public/close.svg';

import classnames from 'classnames';

import { GENRES } from '../../../consts/const';

import styles from '../../../../styles/EditMovieModal.module.scss'


import { MultiSelectField } from '../components/MultiSelectField/MultiSelectField';

// import { getStartButton } from '../../../store/selectors';
import { validationSchema } from '../AddMovieModal/config';

// import { editMovie } from '../../../store/actions';

import { getCorrectValue } from './utils/getCorrectValue';


const EditMovieModalComponent = ({
  onCancel,
  value,
})=> {
  const correctValue = getCorrectValue(value);
  const [state, setState] = useState(correctValue.genres);

  const dispatch = useDispatch();

  const { activeButtonStart, sortStart } = useSelector(getStartButton);

  const handleClick = (item) => {
    setState([...item]);
  };

  const handleSubmitForm = (values, { setSubmitting }) => {
    const correctValues = Object.assign(values, { genres: state });
    correctValues.runtime = +correctValues.runtime;
    editMovie(correctValues, activeButtonStart, sortStart, dispatch);
    setSubmitting(false);
    onCancel();
  };
  return (
    <div className={styles.modal}>
      <Close onClick={onCancel} testId="svg" />
      <p>EDIT MOVIE</p>

      <Formik
        initialValues={correctValue}
        onSubmit={handleSubmitForm}
        validationSchema={validationSchema}
      >
        {({ errors }) => {
          return (
            <Form>
              <InputField
                value={correctValue.id.toString()}
                label="MOVIE ID"
                labelClassName={styles.label}
                className={classnames(styles.field, styles.movieId)}
                disabled
              />
              <InputField
                label="TITLE"
                labelClassName={styles.label}
                className={styles.field}
                name="title"
                id="title"
                error={errors.title}
              />
              <InputField
                label="RELEASE DATA"
                labelClassName={styles.label}
                className={styles.field}
                type="date"
                name="release_date"
                id="release_date"
                error={errors.release_date}
              />
              <InputField
                label="MOVIE URL"
                labelClassName={styles.label}
                className={styles.field}
                name="poster_path"
                id="poster_path"
                error={errors.poster_path}
              />
              <MultiSelectField
                options={GENRES}
                value={state}
                onChange={handleClick}
                placeholder="Select Genre"
                className={styles.multiSelect}
                label="GENRE"
                labelClassName={styles.label}
              />
              <InputField
                label="OVERVIEW"
                labelClassName={styles.label}
                className={styles.field}
                name="overview"
                id="overview"
              />
              <InputField
                label="RUNTIME"
                labelClassName={styles.label}
                className={styles.field}
                name="runtime"
                id="runtime"
              />
              <div className={styles.buttons}>
                <Button className={styles.button} type="reset">
                  RESET
                </Button>
                <Button className={styles.button} type="submit">
                  SUBMIT
                </Button>
              </div>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};

export const EditMovieModal = ({ value }) => {
  const [isOpen, setOpen] = useState(true);

  const handleClose = useCallback(() => {
    setOpen(!isOpen);
  }, [isOpen]);

  return (
    <Modal open={isOpen} onClose={handleClose}>
      <EditMovieModalComponent onCancel={handleClose} value={value} />
    </Modal>
  );
};
