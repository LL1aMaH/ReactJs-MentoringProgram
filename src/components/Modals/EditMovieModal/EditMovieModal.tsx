import React, { memo, useCallback, useState } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import { Formik, FormikHelpers, Form } from 'formik';

import { InputField, EType } from 'Components/InputField';
import { Button } from 'Components/Button';
import { Modal } from 'Components/Modal';

import Close from 'Assets/icons/close.svg';

import classnames from 'classnames';

import { GENRES } from 'Consts/index';
import { TMovie } from '../../../store/types';

import { MultiSelectField } from '../components/MultiSelectField';
import styles from './EditMovieModal.css';

import { getStartButton } from '../../../store/selectors';
import { validationSchema } from '../AddMovieModal/config';

import { editMovie } from '../../../store/actions';

import { getCorrectValue } from './utils';

export type TEditMovieModalComponentProps = {
  onCancel: () => void;
  value: TMovie;
};

export type TEditMovieModalProps = {
  value: TMovie;
};

const EditMovieModalComponent = memo(function EditMovieModalComponent({
  onCancel,
  value,
}: TEditMovieModalComponentProps): JSX.Element {
  const correctValue = getCorrectValue(value);
  const [state, setState] = useState<string[]>(correctValue.genres);

  const dispatch = useDispatch();

  const { activeButtonStart, sortStart } = useSelector(getStartButton);

  const handleClick = (item: string[]) => {
    setState([...item]);
  };

  const handleSubmitForm = (values: TMovie, { setSubmitting }: FormikHelpers<TMovie>) => {
    const correctValues = Object.assign(values, { genres: state });
    correctValues.runtime = +correctValues.runtime;
    dispatch(editMovie(correctValues, activeButtonStart, sortStart));
    setSubmitting(false);
    onCancel();
  };
  return (
    <div className={styles.modal}>
      <Close onClick={onCancel} />
      <p>EDIT MOVIE</p>

      <Formik
        initialValues={correctValue}
        onSubmit={handleSubmitForm}
        validationSchema={validationSchema}
      >
        {({ errors }) => {
          return (
            <Form>
              <InputField
                value={correctValue.id.toString()}
                label="MOVIE ID"
                labelClassName={styles.label}
                className={classnames(styles.field, styles.movieId)}
                disabled
              />
              <InputField
                label="TITLE"
                labelClassName={styles.label}
                className={styles.field}
                name="title"
                id="title"
                error={errors.title}
              />
              <InputField
                label="RELEASE DATA"
                labelClassName={styles.label}
                className={styles.field}
                type={EType.date}
                name="release_date"
                id="release_date"
                error={errors.release_date}
              />
              <InputField
                label="MOVIE URL"
                labelClassName={styles.label}
                className={styles.field}
                name="poster_path"
                id="poster_path"
                error={errors.poster_path}
              />
              <MultiSelectField
                options={GENRES}
                value={state}
                onChange={handleClick}
                placeholder="Select Genre"
                className={styles.multiSelect}
                label="GENRE"
                labelClassName={styles.label}
              />
              <InputField
                label="OVERVIEW"
                labelClassName={styles.label}
                className={styles.field}
                name="overview"
                id="overview"
              />
              <InputField
                label="RUNTIME"
                labelClassName={styles.label}
                className={styles.field}
                name="runtime"
                id="runtime"
              />
              <div className={styles.buttons}>
                <Button className={styles.button} type="reset">
                  RESET
                </Button>
                <Button className={styles.button} type="submit">
                  SUBMIT
                </Button>
              </div>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
});

export const EditMovieModal = memo(function EditMovieModal({
  value,
}: TEditMovieModalProps): JSX.Element {
  const [isOpen, setOpen] = useState(true);

  const handleClose = useCallback(() => {
    setOpen(!isOpen);
  }, [isOpen]);

  return (
    <Modal open={isOpen} onClose={handleClose}>
      <EditMovieModalComponent onCancel={handleClose} value={value} />
    </Modal>
  );
});
