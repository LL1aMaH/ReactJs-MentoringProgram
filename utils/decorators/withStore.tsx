import React from 'react';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import originalStore from './configureStore';

type TComponentFunc = () => JSX.Element;

const withStore = (componentFunc: TComponentFunc, initialState = {}): JSX.Element => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares);
  const store = mockStore(initialState);
  return <Provider store={store}>{componentFunc()}</Provider>;
};

export const withOriginalStore = (componentFunc: TComponentFunc): JSX.Element => {
  return <Provider store={originalStore}>{componentFunc()}</Provider>;
};

export default withStore;
