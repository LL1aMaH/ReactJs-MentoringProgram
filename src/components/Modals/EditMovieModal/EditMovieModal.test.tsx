import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import getFormDecorator from '../../../../utils/decorators/withForm';
import withStore from '../../../../utils/decorators/withStore';

import { EditMovieModal } from './EditMovieModal';

describe('EditMovieModal', () => {
  const withForm = getFormDecorator({ initialValues: { test: 'test' } });
  const value = {
    title: 'title',
    release_date: 'date',
    overview: 'overview',
    runtime: 123,
    genres: [],
    poster_path: 'path',
    id: 1,
  };
  it('renders correctly', () => {
    const getComponent = () =>
      render(
        withStore(() => withForm(() => <EditMovieModal value={value} />), {
          genre: 'ALL',
          sort: 'RELEASE DATE',
        }),
      );

    expect(getComponent().asFragment()).toMatchSnapshot();
  });

  it('renders change correctly', () => {
    const valueTest = {
      title: 'title',
      release_date: 'date',
      overview: 'overview',
      runtime: 123,
      genres: [],
      poster_path: 'path',
      id: 1,
    };
    const component = render(
      withStore(() => withForm(() => <EditMovieModal value={valueTest} />), {
        genre: 'ALL',
        sort: 'RELEASE DATE',
      }),
    );
    const { getByText } = component;

    fireEvent.click(getByText('Action'));
    expect(component.asFragment()).toMatchSnapshot();
  });
});
