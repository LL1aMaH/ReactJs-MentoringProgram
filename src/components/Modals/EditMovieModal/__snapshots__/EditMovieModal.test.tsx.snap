// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMovieModal renders change correctly 1`] = `<DocumentFragment />`;

exports[`EditMovieModal renders correctly 1`] = `<DocumentFragment />`;
